stacks-signer/src/stacks_client.rs

```diff
@@ -0,0 +1,115 @@
+pub struct StacksClient {
+    /// The stacker-db session
+    stackerdb_session: StackerDBSession,
+    /// The private key used in all stacks node communications
+    stacks_private_key: StacksPrivateKey,
+    /// A map of a slot ID to last chunk version
+    slot_versions: HashMap<u32, u32>,
+}
+
+impl From<&Config> for StacksClient {
+    fn from(config: &Config) -> Self {
+        Self {
+            stackerdb_session: StackerDBSession::new(
+                config.node_host,
+                config.stackerdb_contract_id.clone(),
+            ),
+            stacks_private_key: config.stacks_private_key,
+            slot_versions: HashMap::new(),
+        }
+    }
+}
+
+impl StacksClient {
+    /// Sends messages to the stacker-db
+    pub fn send_message(
+        &mut self,
+        id: u32,
+        message: Message,
+    ) -> Result<StackerDBChunkAckData, ClientError> {
+        let message_bytes = bincode::serialize(&message)?;
+        let slot_id = slot_id(id, &message.msg);
+
+        loop {
+            let slot_version = *self.slot_versions.entry(slot_id).or_insert(0) + 1;
+            let mut chunk = StackerDBChunkData::new(slot_id, slot_version, message_bytes.clone());
+            chunk.sign(&self.stacks_private_key)?;
+            debug!("Sending a chunk to stackerdb!\n{:?}", chunk.clone());
+            let chunk_ack = self.stackerdb_session.put_chunk(chunk)?;
+            self.slot_versions.insert(slot_id, slot_version);
+
+            if chunk_ack.accepted {
+                debug!("Chunk accepted by stackerdb: {:?}", chunk_ack);
+                return Ok(chunk_ack);
+            } else {
+                warn!("Chunk rejected by stackerdb: {:?}", chunk_ack);
+            }
+            if let Some(reason) = chunk_ack.reason {
+                // TODO: fix this jankiness. Update stackerdb to use an error code mapping instead of just a string
+                // See: https://github.com/stacks-network/stacks-blockchain/issues/3917
+                if reason == "Data for this slot and version already exist" {
+                    warn!("Failed to send message to stackerdb due to wrong version number {}. Incrementing and retrying...", slot_version);
+                } else {
+                    warn!("Failed to send message to stackerdb: {}", reason);
+                    return Err(ClientError::PutChunkRejected(reason));
+                }
+            }
+        }
+    }
+
+    /// Retrieve the total number of slots allocated to a stacker-db writer
+    #[allow(dead_code)]
+    pub fn slots_per_user(&self) -> u32 {
+        // TODO: retrieve this from the stackerdb instance and make it a function of a given signer public key
+        // See: https://github.com/stacks-network/stacks-blockchain/issues/3921
+        SLOTS_PER_USER
+    }
+}
+
+/// Helper function to determine the slot ID for the provided stacker-db writer id and the message type
+fn slot_id(id: u32, message: &MessageTypes) -> u32 {
```

Comment:
Linking the comment I had on the other PR here for visibility: https://github.com/stacks-network/stacks-blockchain/pull/3923#discussion_r1326362766